Specs/GCMV8.lean

```diff
@@ -0,0 +1,262 @@
+    if i < n then
+      have h2 : (n - i - 1) - (n - i - 1) + 1 = 1 := by omega
+      let xi : BitVec 1 := h2 ▸ extractLsb (n - i - 1) (n - i - 1) x
+      have h3 : m = m - 1 - 0 + 1 := by omega
+      let zi : BitVec m := h3 ▸ extractLsb (m - 1) 0 ((GCMV8.reduce y z) ++ xi)
+      have h1 : 1 = GCMV8.degree y - GCMV8.degree y + 1 := by omega
+      let bit : BitVec 1 := h1 ▸ extractLsb (GCMV8.degree y) (GCMV8.degree y) zi
+      have h4 : 1 = (n - i - 1) - (n - i - 1) + 1 := by omega
+      let newacc : BitVec n := BitVec.partInstall (n - i - 1) (n - i - 1) (h4 ▸ bit) acc
+      pdivTR x y (i + 1) zi newacc
+    else acc
+  pdivTR x y 0 (BitVec.zero m) (BitVec.zero n)
+
+example : pdiv 0b1101#4 0b10#2 (by omega) = 0b110#4 := by native_decide
+example : pdiv 0x1a#5 0b10#2 (by omega) = 0b1101#5 := by native_decide
+example : pdiv 0b1#1 0b10#2 (by omega) = 0b0#1 := by native_decide
+
+/-- Performs modulus of polynomials over GF(2). -/
+def pmod (x : BitVec n) (y : BitVec (m + 1)) (H : m > 0) : BitVec m :=
+  let rec pmodTR (x : BitVec n) (y : BitVec (m + 1)) (p : BitVec (m + 1))
+    (i : Nat) (r : BitVec m) (H : m > 0) : BitVec m :=
+    if i < n then
+      let xi := getLsb x i
+      have h : m - 1 + 1 = m := by omega
+      let tmp := if xi then extractLsb (m - 1) 0 p else h ▸ BitVec.zero m
+      let r := h ▸ r ^^^ tmp
+      pmodTR x y (GCMV8.reduce y (p <<< 1)) (i + 1) (h ▸ r) H
+    else r
+  if y = 0 then 0 else pmodTR x y (GCMV8.reduce y 1) 0 (BitVec.zero m) H
+
+example: pmod 0b011#3 0b00#2 (by omega) = 0b0#1 := by native_decide
+example: pmod 0b011#3 0b01#2 (by omega) = 0b0#1 := by native_decide
+example: pmod 0b011#3 0b10#2 (by omega) = 0b1#1 := by native_decide
+example: pmod 0b011#3 0b11#2 (by omega) = 0b0#1 := by native_decide
+example: pmod 0b011#3 0b100#3 (by omega) = 0b11#2 := by native_decide
+example: pmod 0b011#3 0b1001#4 (by omega) = 0b11#3 := by native_decide
+
+------------------------------------------------------------------------------
+-- Functions related to GCM
+-- In the following section, we use the "GHASH field" to note the GF field with
+-- the irreducible polynomial P(x) = <| 1 + x + x^^2 + x^^7 + x^^128 |>.
+-- We use the term "POLYVAL field" to note the GF field with the bit-reflected
+-- irreducible polynomial Q(x) = <| 1 + x^^121 + x^^126 + x^^127 + x^^128 |>.
+
+/-- Binary representation for
+    irreducible polynomial P(x) = <| 1 + x + x^^2 + x^^7 + x^^128 |>
+-/
+def irrepoly : BitVec 129 := 0x100000000000000000000000000000087#129
+
+/-- Binary representation for
+    irreducible polynomial Q(x) = <| 1 + x^^121 + x^^126 + x^^127 + x^^128 |>
+    https://link.springer.com/chapter/10.1007/978-3-031-34671-2_30
+-/
+def refpoly : BitVec 129 := 0x1C2000000000000000000000000000001#129
+
+/-- H_hat : twisted H
+  H_hat = x × H (mod Q(x))
+  See Remark 5 in paper
+    "A New Interpretation for the GHASH Authenticator of AES-GCM"
+-/
+protected def gcm_init_H (H : BitVec 128) : BitVec 128 :=
```

Comment:
What's the rationale behind protecting some of these functions?

Reply:
Usually, I protect a function when I think it's just a helper function and should not be used outside of the module. Let me know if that's what it's supposed to be.

Reply:
Ah, do you mean to use `private` then instead of `protected`?

Reply:
Hmmmmmm. I didn't realize there is also `private`. Let me try.

Reply:
Done.